README.md
# flutter_wechat_ble

![](screen.png)

<p align="center">
    <a href="https://pub.dartlang.org/packages/flutter_wechat_ble">
        <img src="https://img.shields.io/pub/v/flutter_wechat_ble.svg" alt="pub package" />
    </a>
    <a target="_blank" href="https://shang.qq.com/wpa/qunwpa?idkey=a71a2504cda4cc9ace3320f2dc588bdae928abc671e903463caeb71ec9302c2c"><img border="0" src="https://pub.idqqimg.com/wpa/images/group.png" alt="best-flutter" title="best-flutter"></a>
</p>

A ble library with wechat api style for flutter.

微信api风格的蓝牙库


## 安装

加入依赖:

```
flutter_wechat_ble: any
```

导入

```
import 'package:flutter_wechat_ble/flutter_wechat_ble.dart';
```

## 使用

参考下面的文档，因为语言不一样，所以在调用上略有区别:

https://developers.weixin.qq.com/miniprogram/dev/api/wx.readBLECharacteristicValue.html


#### 全局错误码

在使用本库的时候，所有的错误都可以使用如下的形式来捕获


```
try {
  await FlutterWechatBle.someMethod();
} on BleError catch (e) {
   //// 这里捕获了BleError，表示蓝牙操作发生了异常
   print("错误码 ${e.code} 错误信息 ${e.message}");
}
```


具体的错误码含义如下:

```
String get message {
    switch (code) {
      case "10000":
        return "openBluetoothAdapter not called yet!";
      case "10001":
        return "Bluetooth is not opened!";
      case "10002":
        return "Cannot find device id ";
      case "10003":
        return "Connection fail";
      case "10004":
        return "Cannot find service";
      case "10005":
        return "CHARACTERISTIC  not found";
      case "10006":
        return "No connection found";
      case "10007":
        return "Property not support";
      case "10008":
        return "System error!";
    }
  }
```

这里的错误码定义与微信小程序的错误定义一致，也可以参考这里:
https://developers.weixin.qq.com/miniprogram/dev/api/wx.readBLECharacteristicValue.html


#### 集成:

ios需要注意的是在info.plist中增加权限:

```
<key>NSBluetoothPeripheralUsageDescription</key>
<string>请求访问蓝牙</string>
```

### 查看蓝牙是否可用

```
static FlutterWechatBle.getBluetoothAdapterState()
```


#### 打开蓝牙

```
await FlutterWechatBle.openBluetoothAdapter();
```

注意本库并不对权限做出检查。

#### 关闭蓝牙

```
await FlutterWechatBle.closeBluetoothAdapter();
```

#### 扫描外设:

```
await FlutterWechatBle.startBluetoothDevicesDiscovery();
```


#### 监听外设扫描结果

```
 FlutterWechatBle.onBluetoothDeviceFound((BleDevice device){
    //找到了一个外设
 });
```

BleDevice为扫描的结果，包含一个设备的描述信息

```
  /// uuid of the device
  final String deviceId;

  /// device name
  final String name;

  /// RSSI
  final String RSSI;
```


如果不想监听，则调用

```
FlutterWechatBle.onBluetoothDeviceFound(null)
```
将监听函数设置null即可



#### 获取所有发现的设备

发现的设备将在再次调用startBluetoothDevicesDiscovery后被清空

```
await FlutterWechatBle.getBluetoothDevices();
```

#### 获取所有已经连接的设备

```
await FlutterWechatBle.getConnectedBluetoothDevices();
```


#### 停止扫描外设

```
await FlutterWechatBle.stopBluetoothDevicesDiscovery();

```


#### 连接外设

```
await FlutterWechatBle.createBLEConnection(deviceId: device.deviceId)
```

这里的deviceId为上面的监听函数结果BleDevice的一个属性

#### 断开外设

```
await FlutterWechatBle.closeBLEConnection(deviceId: device.deviceId);
```

#### 监听外设连接状态


```
 FlutterWechatBle.onBLEConnectionStateChange((String deviceId,bool connected){

 });
```

停止监听外设连接状态:
```
FlutterWechatBle.onBLEConnectionStateChange(null);
```

#### 获取服务

```
List<BleService> services = await FlutterWechatBle.getBLEDeviceServices(deviceId: device.deviceId);
```

BleService的属性:
```
 /// uuid of the service
  final String uuid;

  /// always true in android and the `isPrimary` field of the class `CBService` in ios
  final bool isPrimary;
```


#### 获取服务的特征

```
 Future<List<BleCharacteristic>> characteristics = await FlutterWechatBle.getBLEDeviceCharacteristics(
                                                                   deviceId: device.deviceId, serviceId: service.uuid));
```
参数分别为:deviceId和serviceId,

serviceId为上面BleService的字段uuid

BleCharacteristic的属性:


```
  /// uuid of the characteristic
  final String uuid;

  /// support read
  final bool read;

  /// support write
  final bool write;

  /// support notify
  final bool notify;

  /// support indicate
  final bool indicate;
```

与小程序略有区别，将属性直接放在BleCharacteristic这个类下面，使用更加方便


#### 修改notify值

```

await FlutterWechatBle.notifyBLECharacteristicValueChange(
              deviceId: device.deviceId,
              serviceId: service.uuid,
              characteristicId: characteristic.uuid,
              state: true)

```

注意这个方法只能用于支持notify的特征调用。


#### 写入值

```
 await FlutterWechatBle.writeBLECharacteristicValue(
        deviceId: device.deviceId,
        serviceId: service.uuid,
        characteristicId: characteristic.uuid,
        value: 写入值的十六进制字符串或者十六进制数组);
```

这里的value可以为List<int> 或者 十六进制字符串,如果对这些概念不熟悉，可以参考这里：

https://baike.baidu.com/item/%E5%8D%81%E5%85%AD%E8%BF%9B%E5%88%B6/4162457?fr=aladdin


#### 读取值

```
BleValue value =  await FlutterWechatBle.readBLECharacteristicValue(
                       deviceId: device.deviceId,
                       serviceId: service.uuid,
                       characteristicId: characteristic.uuid
                     )
```

BleValue的属性:

```
 final String deviceId;
  final String serviceId;
  final String characteristicId;
  final String value;
```

这里的value为十六进制字符串，如果要得到十六进制数组,则可以使用

```
value.bytes
```


#### 监听值改变

```
 FlutterWechatBle.onBLECharacteristicValueChange((BleValue value){

 });
```

如果需要停止监听:

```
FlutterWechatBle.onBLECharacteristicValueChange(null)
```

设置为null即可




## 加入BluetoothService封装一问一答式蓝牙设备的操作

### 介绍

一问一答式的通信方式指的是将蓝牙设备当成一台服务器来看，当app发送一个设备支持的指令给设备的时候，设备将会通过通知的方式返回一个响应。
这个过程，对于大部分蓝牙设备都是通用的。

其中的一问一答式，我们可以将之封装起来，通过设备配置类，封装数据包的组装和处理的方式，大大减少实际的开发工作量。

目前在代码中放了一个实际的设备,具体可以看这里:

https://github.com/best-flutter/flutter_wechat_ble/blob/master/example/lib/screens/connect_to_tkb.dart

https://github.com/best-flutter/flutter_wechat_ble/blob/master/example/lib/services/TkbDeviceConfig.dart

其中将搜索设备、连接设备、发现服务、发现特征、设置通知、写入和获取数据统一封装起来。将这些复杂的操作转成标准的流程，
并且可以支持连接多种设备，管理多个设备的连接。并将通知型的消息获取过程，改造成发送消息后可以直接使用await等待数据返回，
将原来的非线性流程改成如丝滑般柔顺的线性流程。

下面的代码演示:

```
static DeviceConfig config = new TbkDeviceConfig();
  static BluetoothService bluetoothService =
      new BluetoothService(configs: [config]);


@override
  void initState() {
    super.initState();

    this.startup();
  }

  void startup() async {
    await setState(() {
      messages.add("searing devices...");
    });
    await bluetoothService.shutdown();
    bluetoothService.onServiceDeviceFound(onServiceDeviceFound);
    await bluetoothService.startScan();
  }

  void onServiceDeviceFound(BluetoothServiceDevice device) async {
    print("device ${device.device} ${device.name}");
    await setState(() {
      messages.add("startup devices with name TKB_BLE...${device.name}");
    });
    try {
      await bluetoothService.stopScan();
      await bluetoothService.startupDevice(device.deviceId);

      await setState(() {
        messages.add("writing data : 000062");
      });
      print("write data");
      HexValue value = await device.write("000062");
      print("write data success");
      await setState(() {
        messages.add("writing data success, response: ${value.string}");
      });
      print("=================" + value.string);
    } on BleError catch (e) {
      print(
          ">>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> ${e.code} ${e.message}");
      setState(() {
        messages.add("Ble error : ${e.code} ${e.message}");
      });
    } catch (e) {
      print(">>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> $e");
      setState(() {
        messages.add("other error : ${e.code} ${e.message}");
      });
    }
  }


```





## Example

<img src="images/1.jpg" width="240">  <img src="images/2.jpg" width="240">  <img src="images/3.jpg" width="240">

<img src="images/4.jpg" width="240"/>














CHANGELOG.md


# 0.1.4

* Dont clean devices after calling startBluetoothDevicesDiscovery

## 0.1.3

* Fix connection bug in android

## 0.1.2
* Fix HexUtils.encodeHex bug

## 0.1.1

* Remove v7


## 0.1.0
* add param BluetoothService service in DeviceConfig.onExtraPack

## 0.0.9
* Add wiriteWithReturnData in BluetoothServiceDevice

## 0.0.8
* Add getBluetoothAdapter in ios


## 0.0.7

* Add reconnect in native code


## 0.0.6

* Add reconnect


## 0.0.5

* Add RSSI
* Add example for BluetoothService


## 0.0.4

* Add example for device TKB_BLE
* Can call openBluetoothAdapter repeatly

## 0.0.3

fix ui thread bug


## 0.0.2

format code

## 0.0.1

android and ios support






lib/flutter_wechat_ble.dart
import 'dart:async';

import 'package:flutter/services.dart';
import 'package:flutter_wechat_ble/utils.dart';

export 'package:flutter_wechat_ble/utils.dart';
export 'package:flutter_wechat_ble/bluetooth_service.dart';

// #define NOT_INIT @"10000"
//    #define NOT_AVALIABLE @"10001"
//    #define NO_DEVICE @"10002"
//    #define CONNECTION_FAIL @"10003"
//    #define NO_SERVICE @"10004"
//    #define NO_CHARACTERISTIC @"10005"
//    #define NO_CONNECTION @"10006"
//    #define PROPERTY_NOT_SUPPOTT @"10007"
//    #define SYSTEM_ERROR @"10008"
//    #define SYSTEM_NOT_SUPPORT @"10009"

class BleError extends Error {
  String code;

  BleError({this.code});

  String get message {
    switch (code) {
      case "10000":
        return "openBluetoothAdapter not called yet!";
      case "10001":
        return "Bluetooth is not opened!";
      case "10002":
        return "Cannot find device id ";
      case "10003":
        return "Connection fail";
      case "10004":
        return "Cannot find service";
      case "10005":
        return "CHARACTERISTIC  not found";
      case "10006":
        return "No connection found";
      case "10007":
        return "Property not support";
      case "10008":
        return "System error!";
    }
    return "System not support";
  }
}

class BleDevice {
  /// uuid of the device
  final String deviceId;

  /// device name
  final String name;

  /// RSSI
  final int RSSI;

  List<BleService> services;

  BleDevice({this.deviceId, this.name, this.RSSI});

  void setServices(List<BleService> services) {
    this.services = services;
  }

  BleService getService(String serviceId) {
    return services
        ?.firstWhere((BleService service) => serviceId == service.uuid);
  }
}

class BleService {
  /// uuid of the service
  final String uuid;

  /// always true in android and the `isPrimary` field of the class `CBService` in ios
  final bool isPrimary;

  List<BleCharacteristic> characteristics;

  BleService({this.uuid, this.isPrimary});

  void setCharacteristic(List<BleCharacteristic> characteristics) {
    this.characteristics = characteristics;
  }

  BleCharacteristic getCharacteristic(String characteristicId) {
    return characteristics?.firstWhere((BleCharacteristic characteristic) =>
        characteristicId == characteristic.uuid);
  }
}

class BleValue {
  BleValue({this.deviceId, this.serviceId, this.characteristicId, this.value});

  factory BleValue.fromMap(data) {
    String deviceId = data['deviceId'];
    String serviceId = data['serviceId'];
    String characteristicId = data['characteristicId'];
    String value = data['value'];
    BleValue bleValue = new BleValue(
        deviceId: deviceId,
        serviceId: serviceId,
        characteristicId: characteristicId,
        value: value);
    return bleValue;
  }

  final String deviceId;
  final String serviceId;
  final String characteristicId;
  final String value;

  List<int> _bytes;

  List<int> get bytes {
    if (_bytes == null) {
      if (value == null) {
        return null;
      }
      _bytes = HexUtils.decodeHex(value);
    }
    return _bytes;
  }

  @override
  String toString() {
    return value;
  }
}

class BleCharacteristic {
  /// uuid of the characteristic
  final String uuid;

  /// support read
  final bool read;

  /// support write
  final bool write;

  /// support notify
  final bool notify;

  /// support indicate
  final bool indicate;

  // this property is valid only if notify = true
  bool active = false;

  BleCharacteristic(
      {this.uuid, this.read, this.write, this.notify, this.indicate});
}

typedef void FoundDeviceCallback(BleDevice device);
typedef void ValueChangeCallback(BleValue value);
typedef void ConnectionStateChangeCallback(String deviceId, bool connected);

class BluetoothState {
  // is bluetooth adapter available?
  // 蓝牙适配器是否可用
  bool available;

  // is searching devices
  // 是否正在搜索设备
  bool discovering;

  BluetoothState({this.available, this.discovering});
}

class FlutterWechatBle {
  static const String code = "code";
  static const MethodChannel _channel =
      const MethodChannel('flutter_wechat_ble');

  static bool allowDuplicatesKey;

  static List<String> services;

  static int interval;
  // static StreamController<BleDevice> _foundDeviceController = new StreamController.broadcast();

  /// we must make sure, same deviceId is not dup
  static Map<String, BleDevice> _devices = new Map();

  static Map<String, BleDevice> _connectedDevices = new Map();

  static FoundDeviceCallback _onBluetoothDeviceFoundCallback;

  static Future handler(MethodCall call) {
    String name = call.method;
    var data = call.arguments;
    switch (name) {
      case "foundDevice":
        {
          String deviceId = data['deviceId'];
          _devices.update(deviceId, (BleDevice oldDevice) {
            if (allowDuplicatesKey) {
              if (_onBluetoothDeviceFoundCallback != null) {
                _onBluetoothDeviceFoundCallback(oldDevice);
              }
            }

            /// update the RSSI

            return oldDevice;
          }, ifAbsent: () {
            BleDevice device = new BleDevice(
                name: data['name'],
                deviceId: data['deviceId'],
                RSSI: data['RSSI']);
            if (_onBluetoothDeviceFoundCallback != null) {
              _onBluetoothDeviceFoundCallback(device);
            }
            return device;
          });
        }
        break;
      case "valueUpdate":
        {
          if (_valueChangeCallback != null) {
            _valueChangeCallback(BleValue.fromMap(data));
          }
        }
        break;
      case "stateChange":
        {
          if (_connectionStateChangeCallback != null) {
            _connectionStateChangeCallback(data['deviceId'], data['connected']);
          }
        }
        break;
    }
  }

  static Future openBluetoothAdapter() async {
    _channel.setMethodCallHandler(handler);
    var result = await _channel.invokeMethod('openBluetoothAdapter');
    if (result[code] != null) {
      throw new BleError(code: result[code]);
    }
    return result;
  }

  static Future closeBluetoothAdapter() async {
    var result = await _channel.invokeMethod('closeBluetoothAdapter', {});
    if (result[code] != null) {
      throw new BleError(code: result[code]);
    }
    _connectedDevices.clear();
    _devices.clear();
    return result;
  }

  static Future startBluetoothDevicesDiscovery(
      {bool allowDuplicatesKey = false,
      List<String> services = const <String>[],
      int interval = 0}) async {
    FlutterWechatBle.allowDuplicatesKey = allowDuplicatesKey;
    FlutterWechatBle.services = services;
    FlutterWechatBle.interval = interval;
    var result =
        await _channel.invokeMethod('startBluetoothDevicesDiscovery', {});
    if (result[code] != null) {
      throw new BleError(code: result[code]);
    }
    return result;
  }

  static Future stopBluetoothDevicesDiscovery() async {
    var result =
        await _channel.invokeMethod('stopBluetoothDevicesDiscovery', {});
    if (result[code] != null) {
      throw new BleError(code: result[code]);
    }
    return result;
  }

  static Future createBLEConnection({String deviceId}) async {
    var result = await _channel
        .invokeMethod('createBLEConnection', {"deviceId": deviceId});
    if (result[code] != null) {
      throw new BleError(code: result[code]);
    }
    _connectedDevices[deviceId] = _devices[deviceId];
    return result;
  }

  static Future closeBLEConnection({String deviceId}) async {
    assert(deviceId != null);
    _connectedDevices.remove(deviceId);
    var result = await _channel
        .invokeMethod('closeBLEConnection', {"deviceId": deviceId});
    if (result[code] != null) {
      throw new BleError(code: result[code]);
    }

    return result;
  }

  static Future<List<BleDevice>> getBluetoothDevices() async {
    return _devices.values.toList();
  }

  static Future<List<BleDevice>> getConnectedBluetoothDevices() async {
    return _connectedDevices.values.toList();
  }

  static ValueChangeCallback _valueChangeCallback;

  static ConnectionStateChangeCallback _connectionStateChangeCallback;

  static void onBLECharacteristicValueChange(ValueChangeCallback callback) {
    _valueChangeCallback = callback;
  }

  static void onBLEConnectionStateChange(
      ConnectionStateChangeCallback callback) {
    _connectionStateChangeCallback = callback;
  }

  static Future<List<BleService>> getBLEDeviceServices(
      {String deviceId}) async {
    assert(deviceId != null);

    /// we just get services from cache
    var result = await _channel
        .invokeMethod('getBLEDeviceServices', {"deviceId": deviceId});
    if (result[code] != null) {
      throw new BleError(code: result[code]);
    }

    List rawServices = result['services'];
    //services
    List<BleService> services = rawServices
        .map((data) =>
            new BleService(uuid: data['uuid'], isPrimary: data['isPrimary']))
        .toList();

    return services;
  }

  static void onBluetoothDeviceFound(FoundDeviceCallback callback) {
    FlutterWechatBle._onBluetoothDeviceFoundCallback = callback;
  }

  static Future<List<BleCharacteristic>> getBLEDeviceCharacteristics(
      {String deviceId, String serviceId}) async {
    var result = await _channel.invokeMethod('getBLEDeviceCharacteristics',
        {"deviceId": deviceId, "serviceId": serviceId});
    if (result[code] != null) {
      throw new BleError(code: result[code]);
    }

    List rawData = result['characteristics'];
    //services
    List<BleCharacteristic> characteristics = rawData
        .map((data) => new BleCharacteristic(
              uuid: data['uuid'],
              read: data['read'],
              write: data['write'],
              notify: data['notify'],
              indicate: data['indicate'],
            ))
        .toList();

    return characteristics;
  }

  static Future<BleValue> readBLECharacteristicValue({
    String deviceId,
    String serviceId,
    String characteristicId,
  }) async {
    assert(deviceId != null);
    assert(serviceId != null);
    assert(characteristicId != null);
    var result = await _channel.invokeMethod('readBLECharacteristicValue', {
      "deviceId": deviceId,
      "serviceId": serviceId,
      "characteristicId": characteristicId,
    });
    if (result[code] != null) {
      throw new BleError(code: result[code]);
    }
    return new BleValue.fromMap(result);
  }

  static Future writeBLECharacteristicValue(
      {String deviceId,
      String serviceId,
      String characteristicId,

      /// string or List<int>
      dynamic value}) async {
    assert(value != null);
    assert(deviceId != null);
    assert(serviceId != null);
    assert(serviceId != null);
    assert(characteristicId != null);
    if (value is List<int>) {
      value = HexUtils.encodeHex(value);
    } else if (!(value is String)) {
      throw new Exception("value must be List<int> or String of hex");
    }
    var result = await _channel.invokeMethod('writeBLECharacteristicValue', {
      "deviceId": deviceId,
      "serviceId": serviceId,
      "characteristicId": characteristicId,
      "value": value
    });
    if (result[code] != null) {
      throw new BleError(code: result[code]);
    }
    return result;
  }

  static Future<BluetoothState> getBluetoothAdapterState() async {
    var result = await _channel.invokeMethod('getBluetoothAdapterState', {});

    return new BluetoothState(
        available: result['available'], discovering: result['discovering']);
  }

  static Future notifyBLECharacteristicValueChange(
      {String deviceId,
      String serviceId,
      String characteristicId,
      bool state}) async {
    assert(deviceId != null);
    assert(serviceId != null);
    assert(characteristicId != null);
    assert(state != null);
    var result =
        await _channel.invokeMethod('notifyBLECharacteristicValueChange', {
      "deviceId": deviceId,
      "serviceId": serviceId,
      "characteristicId": characteristicId,
      "state": state
    });
    if (result[code] != null) {
      throw new BleError(code: result[code]);
    }
    return result;
  }
}
